webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/batch/PurgeOldInventoryFiles.scala

```diff
@@ -0,0 +1,121 @@
+ * A "Related Module" means a set of sources files including their
+ * documentation that, without modification of the Source Code, enables
+ * supplementary functions or services in addition to those offered by
+ * the Software.
+ *
+ * Rudder is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+ *
+ *************************************************************************************
+ */
+
+package com.normation.rudder.batch
+
+import better.files.File
+import com.normation.errors.IOResult
+import com.normation.inventory.domain.InventoryProcessingLogger
+import com.normation.utils.CronParser._
+import com.normation.zio._
+import cron4s.CronExpr
+import java.time.Instant
+import zio._
+import zio.clock.Clock
+import zio.duration._
+
+/**
+ * A scheduler which delete old inventory files under /var/rudder/inventories/{received, failed}
+ * It use a cron-like config parsed by cron4s.
+ */
+class PurgeOldInventoryFiles(
+    optCron:          Option[CronExpr],
+    maxAge:           Duration,
+    cleanDirectories: List[File]
+) {
+
+  val logger = InventoryProcessingLogger
+
+  // check is a file is older than max age (true) or not. Only true for regular file.
+  // maxAge in seconds.
+  def isOlderThanMaxAge(f: File, maxAge: Long, now: Instant): IOResult[Boolean] = {
+    IOResult.effect(f.isRegularFile && f.attributes().creationTime().toInstant.isBefore(now.minusSeconds(maxAge)))
+  }
+
+  // must not fail, will be in a cron
+  val cleanOldFiles: UIO[Unit] = for {
+    t0 <- currentTimeMillis
+    now = Instant.ofEpochMilli(t0)
+    _  <- ZIO
+            .foreach_(cleanDirectories) { dir =>
+              ZIO.foreach_(dir.list.to(Iterable)) { f =>
+                isOlderThanMaxAge(f, maxAge.toSeconds, now).flatMap { older =>
+                  if (older) {
+                    InventoryProcessingLogger.info(
+                      s"Deleting inventory file '${f.pathAsString}' (older than ${maxAge.toString})"
+                    ) *>
+                    IOResult
+                      .effect(f.delete())
+                      .catchAll(err =>
+                        InventoryProcessingLogger.error(s"Error when trying to clean old inventory file '${f.pathAsString}'")
+                      )
+                  } else {
+                    InventoryProcessingLogger.trace(
+                      s"Keeping inventory file '${f.pathAsString}' (younger than ${maxAge.toString})"
+                    )
+                  }
+                }
+              }
+            }
+            .catchAll(err => logger.error(s"Error when cleaning-up old inventory files: ${err.fullMsg}"))
```

Comment:
catching here means: if a directory is not readable (for example), others will never been tried, right?

Reply:
yes, the the catch-all of last ressort. Given the soft convention, it should not be reached appart when things are very broken. I'm not sure that a per-directory handling in addition would be better, perhaps (ie: even after the fact and trying to answer your question, I'm not sure it matters since we will need the outer one in all case).